src/lib/mina_net2/bitswap_block.ml

```diff
@@ -154,73 +176,96 @@ let parse_block block =
-      | Some data ->
-          return data
-      | None ->
-          Or_error.error_string "required block not found"
-    in
-    let%bind links, data = parse_block raw_block_data in
-    let%map tail = or_error_list_bind links ~f:parse_chunks in
-    data :: tail
+  let links = Queue.of_list [ root_hash ] in
+  let chunks = Queue.create () in
+  let%map.Or_error () =
+    with_return (fun { return } ->
+        while Queue.length links > 0 do
+          let hash = Queue.dequeue_exn links in
+          let block =
+            match Map.find blocks hash with
+            | None ->
+                return (Or_error.error_string "required block not found")
+            | Some data ->
+                data
+          in
+          let successive_links, chunk =
+            match parse_block block with
+            | Error error ->
+                return (Error error)
+            | Ok x ->
+                x
+          in
+          List.iter successive_links ~f:(Queue.enqueue links) ;
+          Queue.enqueue chunks chunk
+        done ;
+        Ok ())
   in
-  let%map chunks = parse_chunks root_hash in
-  let total_data_size = List.sum (module Int) chunks ~f:Bigstring.length in
+  let total_data_size = Queue.sum (module Int) chunks ~f:Bigstring.length in
   let data = Bigstring.create total_data_size in
   ignore
-    ( List.fold_left chunks ~init:0 ~f:(fun i chunk ->
-          Bigstring.blit ~src:chunk ~src_pos:0 ~dst:data ~dst_pos:i
+    ( Queue.fold chunks ~init:0 ~f:(fun dst_pos chunk ->
+          Bigstring.blit ~src:chunk ~src_pos:0 ~dst:data ~dst_pos
             ~len:(Bigstring.length chunk) ;
-          i + Bigstring.length chunk)
+          dst_pos + Bigstring.length chunk)
       : int ) ;
   data
 
 let%test_module "bitswap blocks" =
   ( module struct
     let schema_of_blocks ~max_block_size blocks root_hash =
-      let total_blocks = Map.length blocks in
-      let full_link_blocks = ref 0 in
-      let non_max_link_block_count = ref None in
-      let last_block_data_size = ref 0 in
+      let num_total_blocks = Map.length blocks in
+      let num_full_branch_blocks = ref 0 in
+      let num_links_in_partial_branch_block = ref None in
+      let last_leaf_block_data_size = ref 0 in
       let max_links_per_block = max_links_per_block ~max_block_size in
       let rec crawl hash =
         let block = Map.find_exn blocks hash in
         let links, chunk = Or_error.ok_exn (parse_block block) in
-        let num_links = List.length links in
-        last_block_data_size := Bigstring.length chunk ;
-        ( if num_links > 0 then
-          if num_links = max_links_per_block then incr full_link_blocks
-          else
-            match !non_max_link_block_count with
+        ( match List.length links with
+        | 0 ->
+            let size = Bigstring.length chunk in
+            last_leaf_block_data_size :=
+              if !last_leaf_block_data_size = 0 then size
+              else min !last_leaf_block_data_size size
```

Comment:
I understand the logic here, like leaf is the smallest block and for it queue is empty. But use of `min` strikes my sight, could we get rid of it?

Reply:
Although it's just a test and isn't so important.

Reply:
Yeah, the min was just the easy way to make this work, since the "last leaf" isn't the last block you traverse here.